### DISCUSSION TITLE
[SSR] Native `HTMLElement` support

### DISCUSSION THREAD

**Discussion by thescientist13** [lit/lit#2092] [Q&A] [answered]
# Description

I have been exploring the demos and testing them out and so far things are going pretty well!  I wanted to validate that I could also just use native `HTMLElement` with **lit-labs/ssr** package but have been getting mixed results.

# Steps to Reproduce

In either of the demo's _module.js_ files

1. Recreated the `<simple-greeting>` example component from the [lit.dev website](https://lit.dev/playground/#sample=examples/hello-world-javascript) as a native `HTMLElement`
    ```js
    const greetingTemplate = document.createElement('template');
    greetingTemplate.innerHTML = `
      <style>
        :host p {
          color: blue; 
        }
      </style>
      <p>Hello, from Simple Greeting component!</p>
    `;

    export class SimpleGreeting extends HTMLElement {
      constructor() {
        super();

        const shadowRoot = this.attachShadow({mode: 'closed'});
        shadowRoot.appendChild(greetingTemplate.content.cloneNode(true));
      }
    }

    customElements.define('simple-greeting', SimpleGreeting);
    ```
1. And then Included it in the shared template
    ```js
    export const template = (data) =>
      html`
        ${header(data.name)}
        <p>${data.message}</p>
        <h4>repeating:</h4>
        <div>${data.items.map((item, i) => html` <p>${i}) ${item}</p> `)}</div>
        ${Array(3)
          .fill(1)
          .map(
            (_item, i) => html`
              <my-element
                ?wasUpdated=${data.wasUpdated}
                .prop=${`${data.prop}: ${i}`}
                attr=${`${data.attr}: ${i}`}
              ></my-element>
            `
          )}
        <simple-greeting></simple-greeting>
      `;
    ```

## Live Reproduction Link

Created a branch in my minimal reproduction repo
https://github.com/thescientist13/lit-ssr-demo/pull/2

## Expected Results

Would have expected the following to display on the page (in blue)
```html
<p>Hello, from Simple Greeting component!</p>
```

## Actual Results

On initial load, the custom element is in the HTML and wrapped in lit comment parts
```html

<!--lit-part j/WQaEoxqeA=-->
      <header>I'm a my-element!</header>
      <div><i>this.prop</i>: <!--lit-part-->prop-value: 2<!--/lit-part--></div>
      <div><i>this.attr</i>: <!--lit-part-->attr-value: 2<!--/lit-part--></div>
    <!--/lit-part--></template></my-element>
        <!--/lit-part--><!--/lit-part-->
    <simple-greeting></simple-greeting>
  <!--/lit-part-->
```

but it is not displaying its contents on the page
![Screen Shot 2021-08-24 at 12 19 34 PM](https://user-images.githubusercontent.com/895923/130654241-c9cb7223-5fcd-43f4-9cb6-784ae68bfb9d.png)

BUT, when I click the hydrate button, it appears! ✨ 
![Screen Shot 2021-08-24 at 12 20 20 PM](https://user-images.githubusercontent.com/895923/130654242-eeeecaee-2463-477a-93f1-bf5faf4a61eb.png)

## Browsers Affected

- [x] Chrome (haven't checked the others)
- [ ] Firefox
- [ ] Edge
- [ ] Safari 11
- [ ] Safari 10
- [ ] IE 11

**Comment by thescientist13** [comment]
Awesome, discussions is now open, so this could definitely be moved there.  👍

**Comment by kevinpschaaf** [comment, marked as answer]
Hi! Give SSR is still in labs, the docs aren't great yet on how we see all of this going together... so thanks for bearing with us while we try and get the Lit 2 GA release out, and then we'll be able to focus more on getting better SSR docs and API's up as that graduates out of labs.

The answer to your question is no, that won't work out of the box. Lit's SSR implementation abstracts custom element rendering to an `ElementRenderer` interface, which can be implemented to do custom rendering for different web components base classes. We provide this abstraction because Lit SSR by default does not include a spec-compliant DOM shim in node, since we can do Lit rendering much more efficiently than actually creating DOM nodes and re-serializing them back out. Right now, the best writeup of this in on this issue here: https://github.com/webcomponents/community-protocols/issues/7#issuecomment-825151215 (and in [this video](https://www.youtube.com/watch?v=f1j7b696L-E&t=4407s)). 

Lit's SSR only ships with an `ElementRenderer` implementation that handles classes that extend from `LitElement`, which is why your `class extends HTMLElement` is not being rendered.  However, it should be possible to implement an `ElementRenderer` interface and pass them into `render` via `elementRenderers` in the SSR `render`: https://github.com/lit/lit/blob/main/packages/labs/ssr/src/lib/render-lit-html.ts#L506-L519, something like:

```ts
render(template(), {elementRenderers: [LitElementRenderer, MyElementRenderer]});
```

How you actually implement `MyElementRenderer` will depend on what your element does. If it uses DOM API's directly, you're likely to need to pull in more of a [DOM shim](https://github.com/lit/lit/blob/main/packages/labs/ssr/src/lib/dom-shim.ts) than the very minimal one we bring along.

For more information on the `ElementRenderer` interface and what's required, check out the API docs in these:
* [element-renderer.ts](https://github.com/lit/lit/blob/main/packages/labs/ssr/src/lib/element-renderer.ts)
* [lit-element-renderer.ts](https://github.com/lit/lit/blob/main/packages/labs/ssr/src/lib/lit-element-renderer.ts)

Hope that's enough info for now to get you started.

**Reply by thescientist13** [reply]
Ah, great, thanks for all that info.  🙌 

Admittedly I had assumed since `LitElement` was derived from `HTMLElement`, that SSR support would come "for free", as it were.  Wishful thinking on my part then.  😇 

I will dive into those resources though and see what I can come up with, as I think it would be really great to have support for native `HTMLElement` SSR (even if just minimally), just to help showcase that the platform can do whatever the frameworks can do too.  Will probably have some more questions for you soon then.

---

Related, I had also been reviewing [this doc here on Declarative Shadow DOM](https://web.dev/declarative-shadow-dom/) which provides some examples using native `HTMLElement` for creating declarative shadow roots and hydrating from them.  So would it be accurate to say that effectively an `ElementRenderer` is responsible for generating valid ShadowDOM + what ever parts its client side "companion" would need to facilitate hydration?

Like what you see in the HTML from Lit SSR?
```html
<!--lit-part BGKATjnCLx0=-->
<my-element attr="attr-value: 2">
  <!--lit-node 0-->
  <template shadowroot="open">
    <style>
      :host {
        ...
      }
   </style>
   <!--lit-part j/WQaEoxqeA=-->
   <header>I'm a my-element!</header>
   <div><i>this.prop</i>: <!--lit-part-->prop-value: 2<!--/lit-part--></div>
   <div><i>this.attr</i>: <!--lit-part-->attr-value: 2<!--/lit-part--></div>
   <!--/lit-part-->
   </template>
   <!--/lit-part-->
</my-element>
<!--/lit-part-->
```

And so then the DOM shim comes into play in regards to how much DOM you want to support based on the the types of browser API calls you want your custom elements to be able to use, as it relates to what that that initial render state / template will / needs to be?

(Sorry, hopefully I'm getting all the words / concepts right, thanks for all the help so far! 🙏 )

**Comment by kevinpschaaf** [comment]
> So would it be accurate to say that effectively an ElementRenderer is responsible for generative valid ShadowDOM + what ever parts its client side "companion" would need to facilitate hydration?

That's right. `ElementRenderer` creates the strings sent to the client, which would be used for the first paint on the browser. When the custom element definition loads and the element boots up on the client it will already have a pre-parsed ShadowRoot, so it's then the element's responsibility for deciding what to do with the DOM on the server. 

A naive implementation could just blow away the contents and re-create it using the same code it would for a client side render (but this would be inefficient, obviously, and probably an anti-pattern). A more sophisticated element would want to "claim" the server rendered DOM (e.g. re-establish the same references used to update DOM based on state changes as when created via client-side rendering) like in the example you linked :

```js
      // Detect whether we have SSR content already:
      if (this.shadowRoot) {
        // A Declarative Shadow Root exists!
        // wire up event listeners, references, etc.:
        const button = this.shadowRoot.firstElementChild;
        button.addEventListener('click', toggle);
      } else {
        // A Declarative Shadow Root doesn't exist.
        // Create a new shadow root and populate it:
        const shadow = this.attachShadow({mode: 'open'});
        shadow.innerHTML = `<button><slot></slot></button>`;
        shadow.firstChild.addEventListener('click', toggle);
      }
```

In Lit, the comment markers you see in the serialized DOM are used by our hydration code to re-establish references used by data structures in the templating layer (which are anchored around corse-grained "chunks" of static content rather than individual nodes, but that's an implementation detail of Lit's specific hydration approach).

> And so then the DOM shim comes into play in regards to how much DOM you want to support based on the the types of browser API calls you want your custom elements to be able to use, as it relates to what that that initial render state / template will / needs to be?

Yep, exactly right. I'm very interested in creating a fallback `ElementRenderer` that would hopefully be able to work with a wider variety of web components by actually bringing in a full DOM shim, and see how far that would get us.

**Reply by thescientist13** [reply]
Awesome, thanks for confirming!  🌟

**Reply by thescientist13** [reply]
I wonder then if this discussion better falls under the scope of https://github.com/webcomponents/community-protocols/issues/7 instead?  Seems like LitSSR is effectively just an implementation detail of a larger, still be defined, conversation of SSR for Web Components.

**Comment by thescientist13** [comment]
For anyone curious, I started working on a new (and experimental) project called Web Components Compiler (WCC) for facilitating native `HTMLElement` SSR as well as exploring various [community protocols](https://github.com/webcomponents-cg/community-protocols) that are actively being discussed in the Web Components Community Group, (like hydration strategies).  
https://github.com/ProjectEvergreen/wcc

I used it as the demo for my talk at OpenJS World 2022 (Web Components at the Edge), a couple weeks ago showing WCs rendering in Serverless (Architect / AWS) and Edge (Netlify) functions.
https://github.com/thescientist13/web-components-at-the-edge

Just thought I would share, and I definitely have some cool ideas where to go with it and have already had some good conversations with folks I’ve shared it with online, so excited to see what’s next and the integration opportunities possible.  I have an issue for tracking alignment with `ElementRenderer` too, so will hope to start looking into that soon.

Happy to hear your thoughts as well!
